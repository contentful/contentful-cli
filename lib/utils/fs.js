const fs = require('fs')
const { promisify } = require('util')

const accessP = promisify(fs.access)
const mkdirP = promisify(fs.mkdir)

module.exports.writeFileP = promisify(fs.writeFile)
module.exports.readFileP = promisify(fs.readFile)

async function ensureDir(dirPath) {
  try {
    await accessP(dirPath)
  } catch (e) {
    if (e.code !== 'ENOENT') {
      throw e
    }

    await mkdirP(dirPath)
  }
}

module.exports.ensureDir = ensureDir
