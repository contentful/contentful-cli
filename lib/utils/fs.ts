const fs = require('fs')
const { promisify } = require('util')
const { isAbsolute, resolve } = require('path')

const accessP = promisify(fs.access)
const mkdirP = promisify(fs.mkdir)

module.exports.writeFileP = promisify(fs.writeFile)
module.exports.readFileP = promisify(fs.readFile)

module.exports.currentDir = () => process.cwd()

function getPath(path) {
  if (isAbsolute(path)) {
    return path
  }

  return resolve(module.exports.currentDir(), path)
}

module.exports.getPath = getPath

async function ensureDir(dirPath) {
  try {
    await accessP(dirPath)
  } catch (e) {
    if (e.code !== 'ENOENT') {
      throw e
    }

    await mkdirP(dirPath)
  }
}

module.exports.ensureDir = ensureDir
