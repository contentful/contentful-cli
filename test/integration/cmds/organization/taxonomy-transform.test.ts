import { readFileSync } from 'fs'
import nixt from 'nixt'
import { join } from 'path'
const bin = join(__dirname, './../../../../', 'bin')

const organizationId = process.env.CLI_E2E_ORG_ID

const app = () => {
  return nixt({ newlines: true }).cwd(bin).base('./contentful.js ').clone()
}

type Result = {
  stderr: string
  stdout: string
}

const cmd = 'organization taxonomy-transform'

describe('organization taxonomy-transform', () => {
  test('should print help message', done => {
    app()
      .run(`${cmd} --help`)
      .code(0)
      .expect(result => {
        const text = result.stdout.trim()
        expect(text).toMatchSnapshot('taxonomy-transform help data is correct')
      })
      .end(done)
  })

  it('should exit 1 when no args', done => {
    app()
      .run(`${cmd}`)
      .code(1)
      .expect(({ stderr }: Result) => {
        const resultText = stderr.trim()

        expect(resultText).toContain(
          'Usage: contentful organization taxonomy-transform'
        )
      })
      .end(done)
  })

  test('should return transformed concepts / concepts scheme', done => {
    app()
      .run(
        `${cmd} --organization-id ${organizationId} -t ../test/integration/cmds/organization/example-load.js --save-file false`
      )
      .expect(({ stdout }: Result) => {
        const resultText = stdout.trim()

        expect(resultText).toContain('Exporting Concepts')
        expect(resultText).toContain('Exporting Concept Schemes')
        expect(resultText).toContain('Running transform script')
        expect(resultText).toContain('concepts')
        expect(resultText).toContain('conceptSchemes')
      })
      .end(done)
  })

  it('should suppress any log output when silent is true', done => {
    app()
      .run(
        `${cmd} --organization-id ${organizationId} -t ../test/integration/cmds/organization/example-load.js --silent -o output.json`
      )
      .code(0)
      .expect(({ stdout }: Result) => {
        const resultText = stdout.trim()

        expect(resultText).not.toContain('Exporting Concepts')
        expect(resultText).not.toContain('Exporting Concept Schemes')
        expect(resultText).not.toContain('Running transform script')
        expect(resultText).not.toContain('concepts')
        expect(resultText).not.toContain('conceptSchemes')

        const transformedData = readFileSync(
          join(__dirname, './../../../../bin/', 'output.json'),
          'utf8'
        )
        expect(transformedData).toMatchSnapshot('transformed data is correct')
      })
      .end(done)
  })
})
