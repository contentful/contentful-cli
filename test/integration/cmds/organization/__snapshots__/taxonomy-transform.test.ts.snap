// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`organization taxonomy-transform should print help message: taxonomy-transform help data is correct 1`] = `
"Usage: contentful organization taxonomy-transform

Options:
  -h, --help                Show help                                  [boolean]
  --management-token, --mt  Contentful management API token             [string]
  --organization-id, --oid  ID of Organization with source data
                                                             [string] [required]
  --header, -H              Pass an additional HTTP Header              [string]
  --output-file, -o         Output file. It defaults to ./data/<timestamp>-<orga
                            nization-id>-transformed.json               [string]
  --transform-script, -t    Script used to transform the taxonomy data
                                                             [string] [required]
  --save-file               Save the transformed taxonomies as a json file
                                                       [boolean] [default: true]
  --silent, -s              Suppress any log output   [boolean] [default: false]"
`;

exports[`organization taxonomy-transform should suppress any log output when silent is true: transformed data is correct 1`] = `
"{
  \\"taxonomy\\": {
    \\"concepts\\": [
      {
        \\"sys\\": {
          \\"id\\": \\"Mammals\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Mammals\\"
        }
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Carnivores\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Carnivores\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Mammals\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Beagle\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Beagle\\"
        },
        \\"definition\\": {
          \\"en-US\\": \\"Handling large carnivores, Feeding protocols, Safety procedures\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Carnivores\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Herbivores\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Herbivores\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Mammals\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Elephant\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Elephant\\"
        },
        \\"definition\\": {
          \\"en-US\\": \\"Large animal handling, Medical care for large mammals\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Herbivores\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Giraffe\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Giraffe\\"
        },
        \\"definition\\": {
          \\"en-US\\": \\"Giraffe care techniques, Enclosure management\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Herbivores\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Birds\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Birds\\"
        }
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Aquatic\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Aquatic\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Birds\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Penguin\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Penguin\\"
        },
        \\"definition\\": {
          \\"en-US\\": \\"Cold environment handling, Feeding techniques\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Aquatic\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Reptiles\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Reptiles\\"
        }
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Venomous\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Venomous\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Reptiles\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Snake\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Snake\\"
        },
        \\"definition\\": {
          \\"en-US\\": \\"Safe handling, Venom management\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Venomous\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Non-Venomous\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Non-Venomous\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Reptiles\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Tortoise\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Tortoise\\"
        },
        \\"definition\\": {
          \\"en-US\\": \\"Shell care, Enclosure maintenance\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Non-Venomous\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Fish\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Fish\\"
        }
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Freshwater\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Freshwater\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Fish\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Koi\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Koi\\"
        },
        \\"definition\\": {
          \\"en-US\\": \\"Water quality management, Feeding\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Freshwater\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Amphibians\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Amphibians\\"
        }
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Frogs\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Frogs\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Amphibians\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Poison Dart Frog\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Poison Dart Frog\\"
        },
        \\"definition\\": {
          \\"en-US\\": \\"Safe handling, Enclosure management\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Frogs\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Invertebrates\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Invertebrates\\"
        }
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Arthropods\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Arthropods\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Invertebrates\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      },
      {
        \\"sys\\": {
          \\"id\\": \\"Tarantula\\"
        },
        \\"prefLabel\\": {
          \\"en-US\\": \\"Tarantula\\"
        },
        \\"definition\\": {
          \\"en-US\\": \\"Tarantulas are hairy spiders\\"
        },
        \\"broader\\": [
          {
            \\"sys\\": {
              \\"id\\": \\"Arthropods\\",
              \\"type\\": \\"Link\\",
              \\"linkType\\": \\"TaxonomyConcept\\"
            }
          }
        ]
      }
    ],
    \\"conceptSchemes\\": []
  }
}"
`;
